### PR TITLE
Lattice semantics in new Datalog solver

### PR DESCRIPTION
**Pull request by bdahse**

### CODE REVIEW

**Comment by bdahse** [comment]
I'm running into what looks like a compiler bug in the code generation phase. Running any of the tests on this branch reproduces this stacktrace:
```
java.lang.IllegalStateException: Result is Err(MismatchedTypes(Denotation['1704],Denotation)).
	at ca.uwaterloo.flix.util.Result.get(Result.scala:36)
	at ca.uwaterloo.flix.util.Result.get$(Result.scala:34)
	at ca.uwaterloo.flix.util.Result$Err.get(Result.scala:74)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.$anonfun$getTagsOf$1(JvmOps.scala:746)
	at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:646)
	at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:642)
	at scala.collection.AbstractIterable.foldLeft(Iterable.scala:919)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.getTagsOf(JvmOps.scala:743)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.nestedTypesOf(JvmOps.scala:1047)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.$anonfun$typesOf$18(JvmOps.scala:1012)
	at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:117)
	at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:104)
	at scala.collection.immutable.HashSet.flatMap(HashSet.scala:34)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.typesOf(JvmOps.scala:1012)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.tagsOf(JvmOps.scala:809)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$2(JvmBackend.scala:72)
	at ca.uwaterloo.flix.api.Flix.subphase(Flix.scala:441)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$1(JvmBackend.scala:51)
	at ca.uwaterloo.flix.api.Flix.phase(Flix.scala:402)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.run(JvmBackend.scala:43)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.run(JvmBackend.scala:33)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.$anonfun$run$1(Phase.scala:42)
	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.api.Flix.codeGen(Flix.scala:376)
	at ca.uwaterloo.flix.api.Flix.$anonfun$compile$1(Flix.scala:390)
	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
	at ca.uwaterloo.flix.api.Flix.compile(Flix.scala:389)
```

I'm currently blocked by this and have not been able to find a workaround.

**Comment by magnus-madsen** [comment]
> 
> 
> I'm running into what looks like a compiler bug in the code generation phase. Running any of the tests on this branch reproduces this stacktrace:
> 
> ```
> java.lang.IllegalStateException: Result is Err(MismatchedTypes(Denotation['1704],Denotation)).
> 	at ca.uwaterloo.flix.util.Result.get(Result.scala:36)
> 	at ca.uwaterloo.flix.util.Result.get$(Result.scala:34)
> 	at ca.uwaterloo.flix.util.Result$Err.get(Result.scala:74)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.$anonfun$getTagsOf$1(JvmOps.scala:746)
> 	at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:646)
> 	at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:642)
> 	at scala.collection.AbstractIterable.foldLeft(Iterable.scala:919)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.getTagsOf(JvmOps.scala:743)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.nestedTypesOf(JvmOps.scala:1047)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.$anonfun$typesOf$18(JvmOps.scala:1012)
> 	at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:117)
> 	at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:104)
> 	at scala.collection.immutable.HashSet.flatMap(HashSet.scala:34)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.typesOf(JvmOps.scala:1012)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.tagsOf(JvmOps.scala:809)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$2(JvmBackend.scala:72)
> 	at ca.uwaterloo.flix.api.Flix.subphase(Flix.scala:441)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$1(JvmBackend.scala:51)
> 	at ca.uwaterloo.flix.api.Flix.phase(Flix.scala:402)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.run(JvmBackend.scala:43)
> 	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.run(JvmBackend.scala:33)
> 	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.$anonfun$run$1(Phase.scala:42)
> 	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
> 	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
> 	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
> 	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
> 	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
> 	at ca.uwaterloo.flix.api.Flix.codeGen(Flix.scala:376)
> 	at ca.uwaterloo.flix.api.Flix.$anonfun$compile$1(Flix.scala:390)
> 	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
> 	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
> 	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
> 	at ca.uwaterloo.flix.api.Flix.compile(Flix.scala:389)
> ```
> 
> I'm currently blocked by this and have not been able to find a workaround.

Is this when you try to use lattices? Because my part is not done yet

**Comment by magnus-madsen** [comment]
Or is this completely independent?

**Comment by bdahse** [comment]
> Or is this completely independent?

Yeah you shouldn't be able to run any test at all with this, because it pulls in my solver in Flix.scala and crashes the compiler before it can run anything.

**Comment by magnus-madsen** [comment]
I sent you a zoom link.

**Comment by bdahse** [comment]
> 
> 
> I sent you a zoom link.

Alright I'll be there in 5.

**Comment by magnus-madsen** [comment]
I pushed a new commit that should fix it.

**Comment by bdahse** [comment]
> 
> 
> I pushed a new commit that should fix it.

I'll see if it helps, my PC is building in IntelliJ right now so Zoom is out of the question atm :)

**Comment by magnus-madsen** [comment]
That is fine, I think this should fix it.

You should able to write a manual test where you construct the AST yourself.

**Comment by bdahse** [comment]
> That is fine, I think this should fix it.
> 
> You should able to write a manual test where you construct the AST yourself.

Hmmm my own tests that I'm using for development are manual and they crash in the same way.

**Comment by bdahse** [comment]
> > > That is fine, I think this should fix it.
> > > You should able to write a manual test where you construct the AST yourself.
> > 
> > 
> > Hmmm my own tests that I'm using for development are manual and they crash in the same way.
> 
> The crash indicates that in one place you have something like `Denotation` and in another `Denotation[v]`. So somewhere you are "forgetting" to pass a type parameter to `Denotation`.
> 
> What does your example look like?

```
def main(_args: Array[String]): Int32 & Impure =
    /// R(23).
    /// R(23) :- R(23).
    let fact1 = Constraint(HeadAtom(PredSym("R", Unknown),
                                    Relational,
                                    [HeadTerm.Lit(23, Unknown)], Unknown),
                            [], Unknown);
    let rule1 = Constraint(HeadAtom(PredSym("R", Unknown),
                                    Relational,
                                    [HeadTerm.Lit(23, Unknown)], Unknown),
                            [BodyAtom(PredSym("R", Unknown),
                                        Relational,
                                        Polarity.Positive,
                                        [BodyTerm.Lit(23, Unknown)], Unknown)],
                            Unknown);
    solve(Datalog([fact1], [rule1]));
    println("hello world!");
    0
```

If it's an ommision on my part, then why am I not getting a type error? It makes sense if the type error is in the Scala code, but not really if it's in Flix code?

**Comment by magnus-madsen** [comment]
Can you post your imports too? A complete program I can run

**Comment by bdahse** [comment]
> Can you post your imports too? A complete program I can run

```
use Fixpoint.{solveWithStratification, solve};
use Fixpoint/Ast.{HeadTerm, BodyTerm, Polarity};
use Fixpoint/Ast.Constraint.Constraint;
use Fixpoint/Ast.Datalog.Datalog;
use Fixpoint/Ast.HeadPredicate.HeadAtom;
use Fixpoint/Ast.BodyPredicate.BodyAtom;
use Fixpoint/Ast.PredSym.PredSym;
use Fixpoint/Ast.VarSym.VarSym;
use Fixpoint/Ast.SourceLocation.Unknown;
use Fixpoint/Ast.Denotation.{Relational};

def main(_args: Array[String]): Int32 & Impure =
    /// R(23).
    /// R(23) :- R(23).
    let fact1 = Constraint(HeadAtom(PredSym("R", Unknown),
                                    Relational,
                                    [HeadTerm.Lit(23, Unknown)], Unknown),
                            [], Unknown);
    let rule1 = Constraint(HeadAtom(PredSym("R", Unknown),
                                    Relational,
                                    [HeadTerm.Lit(23, Unknown)], Unknown),
                            [BodyAtom(PredSym("R", Unknown),
                                        Relational,
                                        Polarity.Positive,
                                        [BodyTerm.Lit(23, Unknown)], Unknown)],
                            Unknown);
    solve(Datalog([fact1], [rule1]));
    println("hello world!");
    0
```

Try this. I haven't run it with your fix but your fix shouldn't matter for this case.

**Comment by bdahse** [comment]
> When I run that I get:
> 
> ```
> ********************************************************************************
> ** Datalog Input
> ********************************************************************************
> 
>     R(23).
>     R(23) :- R(23).
> 
> 
> ********************************************************************************
> ** Relation Algebra Machine AST
> ********************************************************************************
> 
>     // R(23).;
>     project (23) into R;
>     // R(23) :- R(23).;
>     query {0$ ∈ R | 0$[0] = 23} do
>         project (23) into R
>     end;
>     merge R into ΔR;
>     until(ΔR = ∅) do
>         purge ΔR';
>         // R(23) :- R(23).;
>         if((23) ∉ R) then
>             query {0$ ∈ ΔR | 0$[0] = 23} do
>                 project (23) into ΔR'
>             end
>         end;
>         merge ΔR' into R;
>         ΔR := ΔR'
>     end
> 
> 
> ********************************************************************************
> ** Minimal Model
> ********************************************************************************
> 
>     R(23).
> 
> 
> hello world!
> ```

Okay then your change in Lowering.scala does somehow matter because it crashes without it. In any case, that means I can get back to work! :)

File: main/src/library/Fixpoint/Ast/Denotation.flix [hunk 1]
```diff
@@ -15,7 +15,7 @@
  */
 namespace Fixpoint/Ast {
     pub enum Denotation[v] {
-        case Relational,
-        case Latticenal(Unit -> v, v -> v -> v, v -> v -> v, v -> v -> v)
+        case Relational
+        case Latticenal(Unit -> v, v -> v -> Bool, v -> v -> v, v -> v -> v)
```
Comment (bdahse) [inline main/src/library/Fixpoint/Ast/Denotation.flix:19]: I assume the lattice components are given in the order: (bot, leq, lub, glb)

**Comment by bdahse** [comment]
This passes my handcoded lattice test (taken from the Flix website):
```
A("a"; Pos).
B("a"; Top).
A("b"; Neg).
R("c"; s) :- A("a"; s).
R("c"; s) :- A("b"; s).
R("d"; s) :- A(x; s), B(x; s).
```
The overall strategy is:

- Augment RelSym to carry the denotation of the relation it represents.
- Introduce a Meet term that computes the cap of two terms. To compile a rule like `R("d"; s) :- A(x; s), B(x; s).` I unify `s` with the term `Meet(glb, A[1], B[1])` where `glb` is the function from the denotation of `A` and `B` (which is assumed to be the same). Thus all occurences of `s` in the compilation of that rule is substituted for the term  `Meet(glb, A[1], B[1])`.
- Change the database representation from `MutMap[RelSym, MutSet[Array[v]]` to `MutMap[RelSym, MutMap[Array[v], v]`. If a symbol is relational, its tuples are mapped to `default` (and the value is ignored during evaluation).
- Split the environment in two: `type alias SearchEnv[v] = (Array[Tuple[v]], Array[v])`, one to access tuples and one to access lattice elements. A new Lat term is used to distinguish at runtime which environment to look in.
- Use a new Leq expression to compare all latticenal terms, as the latticenal equivalent of Eq used to compare relational terms.

I could compile Meet and Leq into App2 and Guard2 instead. The reason I chose to make new terms and expressions is that it preserves information that are useful for simplification. For example, if I see a `Leq(f, a, b)` expression, I know that `f` is reflexive.

**Comment by magnus-madsen** [comment]
> Introduce a Meet term that computes the cap of two terms. To compile a rule like R("d"; s) :- A(x; s), B(x; s). I unify s with the term Meet(glb, A[1], B[1]) where glb is the function from the denotation of A and B (which is assumed to be the same). Thus all occurences of s in the compilation of that rule is substituted for the term Meet(glb, A[1], B[1]).

Is this appropriately generalized if `s` occurs in three body atoms?

> Change the database representation from MutMap[RelSym, MutSet[Array[v]] to MutMap[RelSym, MutMap[Array[v], v]. If a symbol is relational, its tuples are mapped to default (and the value is ignored during evaluation).

This is OK for now, but probably imposes a performance cost.

> Split the environment in two: type alias SearchEnv[v] = (Array[Tuple[v]], Array[v]), one to access tuples and one to access lattice elements. A new Lat term is used to distinguish at runtime which environment to look in.

Sounds good!

> I could compile Meet and Leq into App2 and Guard2 instead. The reason I chose to make new terms and expressions is that it preserves information useful for simplification. For example, if I see a Leq(f, a, b) expression, I know that f is reflexive.

I agree. I prefer to keep the information.

**Comment by bdahse** [comment]
> 
> 
> > Introduce a Meet term that computes the cap of two terms. To compile a rule like R("d"; s) :- A(x; s), B(x; s). I unify s with the term Meet(glb, A[1], B[1]) where glb is the function from the denotation of A and B (which is assumed to be the same). Thus all occurences of s in the compilation of that rule is substituted for the term Meet(glb, A[1], B[1]).
> 
> Is this appropriately generalized if `s` occurs in three body atoms?
> 
Yes, in that case Meet is nested (and every meet term carries the same glb function which is a bit redundant).

> > Change the database representation from MutMap[RelSym, MutSet[Array[v]] to MutMap[RelSym, MutMap[Array[v], v]. If a symbol is relational, its tuples are mapped to default (and the value is ignored during evaluation).
> 
> This is OK for now, but probably imposes a performance cost.
> 
Why? Isn't a MutSet[k] equivalent to a MutMap[k, Unit] in terms of performance?

> > Split the environment in two: type alias SearchEnv[v] = (Array[Tuple[v]], Array[v]), one to access tuples and one to access lattice elements. A new Lat term is used to distinguish at runtime which environment to look in.
> 
> Sounds good!
> 
This split can actually be avoided if the environment was a single flat array, i.e. `type alias SearchEnv[v] = Array[v]`. Depending on how fast it is to call System.arraycopy, the flat array would probably have better performance because it has only less layer of indirection and no branching.

> > I could compile Meet and Leq into App2 and Guard2 instead. The reason I chose to make new terms and expressions is that it preserves information useful for simplification. For example, if I see a Leq(f, a, b) expression, I know that f is reflexive.
> 
> I agree. I prefer to keep the information.

**Comment by magnus-madsen** [comment]
> Yes, in that case Meet is nested (and every meet term carries the same glb function which is a bit redundant).

Excellent!

> 
> > > Change the database representation from MutMap[RelSym, MutSet[Array[v]] to MutMap[RelSym, MutMap[Array[v], v]. If a symbol is relational, its tuples are mapped to default (and the value is ignored during evaluation).
> > 
> > 
> > This is OK for now, but probably imposes a performance cost.
> 
> Why? Isn't a MutSet[k] equivalent to a MutMap[k, Unit] in terms of performance?

It is. You are right. My only concern is that `default` is somewhat of a "hidden feature".

> This split can actually be avoided if the environment was a single flat array, i.e. `type alias SearchEnv[v] = Array[v]`. Depending on how fast it is to call System.arraycopy, the flat array would probably have better performance because it has only less layer of indirection and no branching.

OK

**Pull request merged**